apis/v1alpha1/opampbridge_webhook.go

```diff
@@ -0,0 +1,115 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v1alpha1
+
+import (
+	"fmt"
+	"strings"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/validation"
+	ctrl "sigs.k8s.io/controller-runtime"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+)
+
+// log is for logging in this package.
+var opampbridgelog = logf.Log.WithName("opampbridge-resource")
+
+func (r *OpAMPBridge) SetupWebhookWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewWebhookManagedBy(mgr).
+		For(r).
+		Complete()
+}
+
+//+kubebuilder:webhook:path=/mutate-opentelemetry-io-v1alpha1-opampbridge,mutating=true,failurePolicy=fail,sideEffects=None,groups=opentelemetry.io,resources=opampbridges,verbs=create;update,versions=v1alpha1,name=mopampbridge.kb.io,admissionReviewVersions=v1
+
+var _ webhook.Defaulter = &OpAMPBridge{}
+
+// Default implements webhook.Defaulter so a webhook will be registered for the type.
+func (r *OpAMPBridge) Default() {
+	opampbridgelog.Info("default", "name", r.Name)
+	if len(r.Spec.UpgradeStrategy) == 0 {
+		r.Spec.UpgradeStrategy = UpgradeStrategyAutomatic
+	}
+
+	if r.Labels == nil {
+		r.Labels = map[string]string{}
+	}
+	if r.Labels["app.kubernetes.io/managed-by"] == "" {
+		r.Labels["app.kubernetes.io/managed-by"] = "opentelemetry-operator"
+	}
+
+	one := int32(1)
+	if r.Spec.Replicas == nil {
+		r.Spec.Replicas = &one
+	}
+}
+
+//+kubebuilder:webhook:path=/validate-opentelemetry-io-v1alpha1-opampbridge,mutating=false,failurePolicy=fail,sideEffects=None,groups=opentelemetry.io,resources=opampbridges,verbs=create;update;delete,versions=v1alpha1,name=vopampbridge.kb.io,admissionReviewVersions=v1
+
+var _ webhook.Validator = &OpAMPBridge{}
+
+// ValidateCreate implements webhook.Validator so a webhook will be registered for the type.
+func (r *OpAMPBridge) ValidateCreate() (admission.Warnings, error) {
+	opampbridgelog.Info("validate create", "name", r.Name)
+	return nil, r.validateCRDSpec()
+}
+
+// ValidateUpdate implements webhook.Validator so a webhook will be registered for the type.
+func (r *OpAMPBridge) ValidateUpdate(old runtime.Object) (admission.Warnings, error) {
+	opampbridgelog.Info("validate update", "name", r.Name)
+	return nil, r.validateCRDSpec()
+}
+
+// ValidateDelete implements webhook.Validator so a webhook will be registered for the type.
+func (r *OpAMPBridge) ValidateDelete() (admission.Warnings, error) {
+	opampbridgelog.Info("validate delete", "name", r.Name)
+	return nil, nil
+}
+
+func (r *OpAMPBridge) validateCRDSpec() error {
+
+	// check required fields
+
+	if len(strings.TrimSpace(r.Spec.Endpoint)) == 0 {
+		return fmt.Errorf("the OpAMP server endpoint is not specified")
+	}
+
+	if len(strings.TrimSpace(r.Spec.Protocol)) == 0 {
+		return fmt.Errorf("the transport for OpAMP server protocol is not specified")
+	}
```

Comment:
Is there a reason to have this separate `Protocol` option here since the endpoint should already include it? I think the `Protocol` is redundant.

Reply:
this was because the bridge accepted both, i am removing its support for this field.